src/quacc/wflow_tools/decorators.py

```diff
@@ -3,19 +3,16 @@
 from __future__ import annotations
 
 from functools import partial, wraps
-from typing import TYPE_CHECKING, TypeVar
+from typing import Any, Callable, TypeVar
 
 from quacc.settings import change_settings_wrap
 
-Job = TypeVar("Job")
-Flow = TypeVar("Flow")
-Subflow = TypeVar("Subflow")
+Job = TypeVar("Job", bound=Callable[..., Any])
+Flow = TypeVar("Flow", bound=Callable[..., Any])
+Subflow = TypeVar("Subflow", bound=Callable[..., Any])
 
-if TYPE_CHECKING:
-    from typing import Any, Callable
 
-
-def job(_func: Callable | None = None, **kwargs) -> Job:
+def job(_func: Job | None = None, **kwargs) -> Job | Callable:
```

Comment:
I am a bit confused about this one and the ones similar to it. The `@job` decorator is generally applied to a plain function (`Callable`). The output of this is what we are defining internally as a `Job` object. So, in this case, why would `_func` be type hinted as `Job`? I think this would be confusing because it looks like `job` applies to a `Job`, but it applies to a plain old Python function. Same comment for the flow and subflow.

Reply:
If the output of the `_func` is the job, we can change the signature of the job as follows:
```python
def job(_func: Callable[..., Job] | None = None, **kwargs) -> Job | Callable:
```
What do you think ?

Reply:
The output of `_func` itself is anything. It can be any function at all. The output of `job(_func)` is a `Job`, by definition.

So, the type hinting here is still a bit unclear to me. The `_func` is any `Callable` (i.e. any plain function), so `Callable[..., Job]` does not seem accurate. It would just be `Callable`.

Reply:
Ah, that clarifies things. The generic type in the code initially led me to believe the input and output have the same type.  Assuming I'm correct, `_func` is likely executed by various workflow engines (`Covalent`, `Dask`, etc.) based on the configuration. Consequently, both `_func` and `job` end up being of type Callable[..., Any].

Here is the new types:

```python
Job = Callable[..., Any]
Flow = Callable[..., Any]
Subflow = Callable[..., Any]


def job(_func: Callable[..., Any] | None = None, **kwargs) -> Job:
...
```

Reply:
Ah, yes! This looks a lot more like what I'd expect!

Technically, some of the workflow engines return a class rather than a function object. For instance, parsl returns a `PythonApp` class, jobflow returns its own `Job` class, Dask returns a `Delayed` class, and so on. But I have just been referring to these all as `Job`. In other words, I have been defining `Job` in a somewhat circular way as "a `Job` is simply whatever is returned by `def job`. 😅 

I guess we can see if mypy is okay with this type hint as-is!

Reply:
No problem with mypy after this change! All the mypy issues in `decorators.py` are resolved.